doc: update `Performance tuning` section (#19530)
vlang/v: doc/docs.md

    * [Structs with reference fields](#structs-with-reference-fields)
    * [sizeof and __offsetof](#sizeof-and-__offsetof)
    * [Limited operator overloading](#limited-operator-overloading)

>     * [Tuning](#tuning)

Reviewer: Please do not break links without need. The old section name was fine.

Author: Doesn't new name suit better for this section since it describes not only performance tweaks, but also memory usage and executable size tweaks?

Reviewer 2: Aren't those all related to performance?  In a way, at least.  Hard to get something to perform better (or even correctly), if it doesn't fit in memory, and executable size directly affects memory as well.

Reviewer: > Doesn't new name suit better for this section since it describes not only performance tweaks, but also memory usage and executable size tweaks?

No, that is part of the problem. `Tuning` is very general. In a sense, every option passed to the compiler can be considered to be tuning. For example `-g` or `-gc none` or `-autofree` or `-trace-calls` or `-compress` or `-profile`, will all affect the produced executable, but their main goal is not improving performance.

- Arguments cannot be changed inside overloads.
- Calling other functions inside operator functions is not allowed (**planned**).

## Tuning

When compiled with `-prod`, V's generated C code usually performs well. However, in specialized
scenarios, additional compiler flags and attributes can further optimize the executable for
performance, memory usage, or size.

> [!WARNING]
> Use these options only after profiling your code and confirming their benefits.

> > Predicting performance without empirical data can be misleading.

Reviewer: I prefer the old section, since it was both prescriptive, and descriptive.
The quote "programmers are notoriously bad at predicting how their programs actually perform" also has a more nuanced meaning than the bland statement "Predicting performance without empirical data can be misleading." .

| `[minify]`               | Performance, Memory usage       | May break binary serialization/reflection         |
| `_likely_/_unlikely_`    | Performance                     | Risk of negative performance impact               |
| `-skip-unused`           | Performance, Compile time, Size | Potential instability                             |
| `-fast-math`             | Performance                     | Risk of incorrect mathematical operations results |
| `-d no_segfault_handler` | Compile time, Size              | Loss of segfault trace                            |
| `-cflags -march=native`  | Performance                     | Risk of reduced CPU compatibility                 |
| `PGO`                    | Performance, Size               | Usage complexity                                  |

### Tuning operations details

#### `[inline]`

You can tag functions with `[inline]`, so the C compiler will try to inline them, which in some
cases, may be beneficial for performance, but may impact the size of your executable.

**When to Use**

- Functions that are called frequently in performance-critical loops.

> - Functions that are called only once.

Reviewer: Why? Functions that are called only once usually do not matter for performance at all.

Reviewer: The overhead of a single function call is tiny, it makes sense to optimize that only when it is comparable to the work a function itself does, and only if that function is called sufficiently frequently.

**When to Avoid**

- Large functions, as it might cause code bloat and actually decrease performance.
- Large functions in `if` expressions - may have negative impact on instructions cache.

#### `[direct_array_access]`

In functions tagged with `[direct_array_access]` the compiler will translate array operations
directly into C array operations - omitting bounds checking. This may save a lot of time in a
function that iterates over an array but at the cost of making the function unsafe - unless the
boundaries will be checked by the user.

**When to Use**

- In tight loops that access array elements, where bounds have been manually verified or you are
sure that the access index will be valid.

**When to Avoid**

> - When safety is a concern and bounds checks are required.

Reviewer: I would say just "everywhere else".
Some people think that they do not need safety, and bounds checking, only to find out, that they have spent hours debugging problems, that would have been simple panics, caught by the bound checks (it has happened several times on just the Discord channels already...)

#### `[packed]`

The `[packed]` attribute can be added to a structure to create an unaligned memory layout,
which decreases the overall memory footprint of the structure.

> > [!WARNING]

Reviewer: Imho use a `**Note**` here, a warning should be reserved for more important things, that can lead to crashes etc.

> and you're willing to sacrifice performance.

**When to Use**

- When memory usage is more critical than performance, e.g., in embedded systems.

**When to Avoid**

- On CPU architectures that do not support unaligned memory access or when high-speed memory access
is needed.

#### `[minify]`

The `[minify]` attribute can be added to a struct, allowing the compiler to reorder the fields in
a way that minimizes internal gaps while maintaining alignment.

> [!WARNING]
> Using the `[minify]` attribute may cause issues with binary serialization or reflection.
> Be mindful of these potential side effects when using this attribute.

**When to Use**

- When you want to minimize memory usage and you're not using binary serialization or reflection.

**When to Avoid**

- When using binary serialization or reflection, as it may cause unexpected behavior.

#### `_likely_/_unlikely_`

`if _likely_(bool expression) {` - hints to the C compiler, that the passed boolean expression is
very likely to be true, so it can generate assembly code, with less chance of branch misprediction.
In the JS backend, that does nothing.

> `if _unlikely_(bool expression) {` similar to `_likely_(x)`, but it hints that the boolean

Reviewer: ```suggestion
`if _unlikely_(bool expression) {` is similar to `_likely_(x)`, but it hints that the boolean
```

expression is highly improbable. In the JS backend, that does nothing.

**When to Use**

- In conditional statements where one branch is clearly more frequently executed than the other.

**When to Avoid**

- When the prediction can be wrong, as it might cause a performance penalty due to branch
misprediction.

#### `-skip-unused`

> This flag tells the V compiler to omit code that is not needed in the final executable to run your
> program correctly. This will remove unneeded `const` arrays allocations and unused functions
> from your code and the `vlib` in the output executable.

Reviewer: ```suggestion
This flag tells the V compiler to omit code that is not needed in the final executable to run your
program correctly. This will remove unneeded `const` arrays allocations and unused functions
from the code in the generated executable.
```

This flag will be on by default in the future when its implementation will be stabilized and all
severe bugs will be found.

**When to Use**

- For production builds where you want to reduce the executable size and improve runtime
performance.

**When to Avoid**

- Where it doesn't work for you.

#### `-fast-math`

This flag enables optimizations that disregard strict compliance with the IEEE standard for
floating-point arithmetic. While this could lead to faster code, it may produce incorrect or
less accurate mathematical results.

> The full specter of math operations that `-fast-math` affects can be found [here](https://clang.llvm.org/docs/UsersManual.html#cmdoption-ffast-math).

Reviewer: ```suggestion
The full specter of math operations that `-fast-math` affects can be found
[here](https://clang.llvm.org/docs/UsersManual.html#cmdoption-ffast-math).
```
that will help `v check-md`, and is easier to read in terminals .

**When to Use**

- In applications where performance is more critical than precision, like certain graphics
rendering tasks.

**When to Avoid**

- In applications requiring strict mathematical accuracy, such as scientific simulations or
financial calculations.

#### `-d no_segfault_handler`

Using this flag omits the segfault handler, reducing the executable size and potentially improving
compile time. However, in the case of a segmentation fault, the output will not contain stack trace
information, making debugging more challenging.

**When to Use**

- In small, well-tested utilities where a stack trace is not essential for debugging.

**When to Avoid**

- In large-scale, complex applications where robust debugging is required.

#### `-cflags -march=native`

> This flag directs the C compiler to generate instructions optimized for the host CPU. This can
> improve performance but may produce an executable incompatible with other/older CPUs.

Reviewer: ```suggestion
This flag directs the C compiler to generate instructions optimized for the host CPU. This can
improve performance but will produce an executable incompatible with other/older CPUs.
```

**When to Use**

- When the software is intended to run only on the build machine or in a controlled environment
with identical hardware.

**When to Avoid**

> - When distributing the software to users with potentially different CPU architectures.

Reviewer: ```suggestion
- When distributing the software to users with potentially older CPUs.
```

Users having CPUs with different architectures, will need an emulator in general.
